File: test/harness.ts
// Builds the installer once, serves fake releases over HTTP, and drives the
// built cli.cjs as a real process with a fake Computer.
import { spawn } from "node:child_process";
import { createHash } from "node:crypto";
import { chmodSync, existsSync, mkdtempSync, mkdirSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { createServer, type Server } from "node:http";
import { tmpdir } from "node:os";
import { join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const root = resolve(fileURLToPath(import.meta.url), "..", "..");
export const fakePath = join(root, "test", "fake-computer.mjs");
const platformKey = `${process.platform}-${process.arch}`;
export const WINDOWS = process.platform === "win32";
/** The fake as a real executable: always on Windows, or RCI_FAKE_SEA=1 anywhere to exercise it. */
const SEA_FAKE = WINDOWS || process.env.RCI_FAKE_SEA === "1";
const BIN = WINDOWS ? "raft-computer.exe" : "raft-computer";
/** RCI_NATIVE=1 drives the single executables instead of node + cli.cjs. */
export const NATIVE = process.env.RCI_NATIVE === "1";

export interface FakeRelease { version: string; startFail?: boolean; reportedVersion?: string; nextStep?: string; stopBroken?: boolean }

export class Harness {
  dist = "";
  server!: Server;
  base = "";
  releases = new Map<string, Buffer>();
  channel: { main?: string; alpha?: string } = {};
  sidecar = Buffer.from("not really wasm but verified all the same");

  fakeSea: Buffer | null = null;
  async start(): Promise<void> {
    this.dist = mkdtempSync(join(tmpdir(), "rci-dist-"));
    await run(process.execPath, [join(root, "scripts", "build.mjs"), this.dist, ...(NATIVE ? ["--native"] : [])], { cwd: root });
    if (SEA_FAKE) {
      const { buildFakeSea } = await import("./build-fake-sea.mjs");
      const { readFileSync } = await import("node:fs");
      this.fakeSea = readFileSync(await buildFakeSea(join(this.dist, "fake"), fakePath));
    }
    this.server = createServer((req, res) => {
      const url = new URL(req.url ?? "/", "http://x");
      const m = /^\/computer\/([^/]+)\/(.+)$/.exec(url.pathname);
      if (m) {
        const [, version, file] = m;
        const bytes = this.releases.get(version);
        if (!bytes) { res.statusCode = 404; return res.end(); }
        if (file === "manifest.json") { res.setHeader("content-type", "application/json"); return res.end(JSON.stringify(this.manifest(version, bytes))); }
        if (file === "raft-computer") return res.end(bytes);
        if (file === "photon_rs_bg.wasm") return res.end(this.sidecar);
      }
      const i = /^\/installer\/(.+)$/.exec(url.pathname);
      if (i) {
        const f = join(this.dist, ...i[1].split("/"));
        if (existsSync(f)) return res.end(readFileSync(f));
        res.statusCode = 404; return res.end();
      }
      if (url.pathname === "/computer/install.sh") return res.end(readFileSync(join(this.dist, "install.sh")));
      const h = /^\/public\/v2\/apps\/([^/]+)\/latest$/.exec(url.pathname);
      if (h) {
        const channel = url.searchParams.get("channel") === "alpha" ? "alpha" : "main";
        const version = this.channel[channel];
        const bytes = version ? this.releases.get(version) : undefined;
        if (!version || !bytes) { res.statusCode = 404; return res.end(); }
        const [platform, arch] = platformKey.split("-");
        return res.end(JSON.stringify({ build: { version }, assets: [{ platform, arch, variant: null, filetype: "binary", size_bytes: bytes.length, sha256: sha(bytes) }] }));
      }
      res.statusCode = 404; res.end();
    });
    await new Promise<void>((r) => this.server.listen(0, "127.0.0.1", r));
    const addr = this.server.address();
    this.base = `http://127.0.0.1:${typeof addr === "object" && addr ? addr.port : 0}`;
  }

  async stop(): Promise<void> {
    this.server.closeAllConnections();
    await new Promise<void>((r) => this.server.close(() => r()));
    rmSync(this.dist, { recursive: true, force: true });
  }

  publish(r: FakeRelease): Buffer {
    const behaviour: Record<string, string> = { RAFT_FAKE_VERSION: r.reportedVersion ?? r.version };
    if (r.startFail) behaviour.RAFT_FAKE_START_FAIL = "1";
    if (r.nextStep) behaviour.RAFT_FAKE_NEXT_STEP = r.nextStep;
    if (r.stopBroken) behaviour.RAFT_FAKE_STOP_BROKEN = "1";
    const bytes = this.fakeSea
      ? Buffer.concat([this.fakeSea, Buffer.from(`\n#RAFT_FAKE:${JSON.stringify(behaviour)}\n`)])
      : Buffer.from(`#!/bin/sh\n${Object.entries(behaviour).map(([k, v]) => `${k}=${JSON.stringify(v)}`).join(" ")} RAFT_FAKE_SELF="$0" exec "${process.execPath}" "${fakePath}" "$@"\n`);
    this.releases.set(r.version, bytes);
    return bytes;
  }

  manifest(version: string, bytes: Buffer) {
    return {
      name: "raft-computer", version,
      photonWasm: { file: "photon_rs_bg.wasm", sha256: sha(this.sidecar), size: this.sidecar.length },
      targets: { [platformKey]: { file: "raft-computer", sha256: sha(bytes), size: bytes.length } },
    };
  }

  machine(): Machine {
    const home = mkdtempSync(join(tmpdir(), "rci-machine-"));
    const installDir = join(home, "bin");
    mkdirSync(installDir, { recursive: true });
    return new Machine(this, home, installDir);
  }
}

export class Machine {
  constructor(readonly h: Harness, readonly home: string, readonly installDir: string) {}
  get binary(): string { return join(this.installDir, BIN); }
  get kStateDir(): string { return join(this.home, "computer", "k"); }
  env(extra: NodeJS.ProcessEnv = {}): NodeJS.ProcessEnv {
    return {
      ...process.env, HOME: this.home, USERPROFILE: this.home, SHELL: "/bin/zsh", RAFT_HOME: this.home, RAFT_COMPUTER_INSTALL_DIR: this.installDir,
      RAFT_COMPUTER_NO_MODIFY_PATH: WINDOWS ? "1" : "0",
      RAFT_COMPUTER_RELEASE_BASE: `${this.h.base}/computer`, RAFT_COMPUTER_HANDS_ORIGIN: this.h.base,
      RAFT_COMPUTER_INSTALLER_RUNNER: NATIVE ? join(this.h.dist, "native", platformKey, WINDOWS ? "raft-computer-installer-runner.exe" : "raft-computer-installer-runner") : join(this.h.dist, "runner.mjs"),
      RAFT_COMPUTER_NON_INTERACTIVE: "1", ...extra,
    };
  }
  /** The published entry point: install.sh (or install.ps1) from the local installer release base. */
  async bootstrap(args: string[], extra: NodeJS.ProcessEnv = {}): Promise<Result> {
    // Portable mode names the Node to use; native mode and a test that
    // wants "no Node at all" (RAFT_COMPUTER_INSTALLER_NODE: "") leave it unset.
    const env = { ...this.env(extra), RAFT_COMPUTER_INSTALLER_RELEASE_BASE: `${this.h.base}/installer` };
    if (!NATIVE && extra.RAFT_COMPUTER_INSTALLER_NODE === undefined) env.RAFT_COMPUTER_INSTALLER_NODE = process.execPath;
    if (!env.RAFT_COMPUTER_INSTALLER_NODE) delete env.RAFT_COMPUTER_INSTALLER_NODE;
    delete env.RAFT_COMPUTER_INSTALLER_RUNNER;
    return WINDOWS
      ? run("powershell.exe", ["-NoProfile", "-ExecutionPolicy", "Bypass", "-File", join(this.h.dist, "install.ps1"), ...args], { env, allowFailure: true })
      : run("/bin/sh", [join(this.h.dist, "install.sh"), ...args], { env, allowFailure: true });
  }
  async installer(args: string[], extra: NodeJS.ProcessEnv = {}): Promise<Result> {
    return NATIVE
      ? run(join(this.h.dist, "native", platformKey, WINDOWS ? "raft-computer-installer.exe" : "raft-computer-installer"), args, { env: this.env(extra), allowFailure: true })
      : run(process.execPath, [join(this.h.dist, "cli.cjs"), ...args], { env: this.env(extra), allowFailure: true });
  }
  /** Put a pre-K Computer on PATH, log in and start it, as the old install.sh plus a user would have. */
  async preinstall(version: string): Promise<void> {
    const bytes = this.h.releases.get(version) ?? this.h.publish({ version });
    writeFileSync(this.binary, bytes, { mode: 0o755 });
    chmodSync(this.binary, 0o755);
    await this.loginAndStart();
  }
  async loginAndStart(): Promise<void> {
    await run(this.binary, ["login"], { env: { ...process.env, RAFT_HOME: this.home } });
    await run(this.binary, ["start"], { env: { ...process.env, RAFT_HOME: this.home } });
  }
  async selfVersion(): Promise<string> {
    return (await run(this.binary, ["--version"], { env: { ...process.env, RAFT_HOME: this.home } })).stdout.trim();
  }
  async live(): Promise<{ version: string; pid: number } | null> {
    const r = await run(this.binary, ["status", "--json"], { env: { ...process.env, RAFT_HOME: this.home }, allowFailure: true });
    if (r.code !== 0) return null;
    const a = JSON.parse(r.stdout).attestation;
    return a ? { version: a.computerVersion, pid: a.servicePid } : null;
  }
  async cleanup(): Promise<void> {
    if (existsSync(this.binary)) {
      try { await run(this.binary, ["stop"], { env: { ...process.env, RAFT_HOME: this.home }, allowFailure: true }); } catch { /* none */ }
    }
    rmSync(this.home, { recursive: true, force: true });
  }
}

export interface Result { code: number; stdout: string; stderr: string }
export function run(cmd: string, args: string[], o: { cwd?: string; env?: NodeJS.ProcessEnv; allowFailure?: boolean } = {}): Promise<Result> {
  return new Promise((resolve, reject) => {
    const p = spawn(cmd, args, { cwd: o.cwd, env: o.env ?? process.env, stdio: ["ignore", "pipe", "pipe"] });
    let stdout = "", stderr = "";
    p.stdout.on("data", (d) => { stdout += d; });
    p.stderr.on("data", (d) => { stderr += d; });
    p.on("error", (error) => reject(error));
    p.on("close", (code) => {
      const r = { code: code ?? 1, stdout, stderr };
      if (r.code === 0 || o.allowFailure) resolve(r); else reject(new Error(`${cmd} ${args.join(" ")} exited ${r.code}\n${stderr}\n${stdout}`));
    });
  });
}
export function sha(b: Buffer): string { return createHash("sha256").update(b).digest("hex"); }

File: verification/harness.py
import gzip
"""Real native processes, isolated homes, loopback release authority and CDN."""
import hashlib
import http.server
import json
import os
from pathlib import Path
import shutil
import socket
import subprocess
import sys
import tempfile
import threading
import time
import urllib.parse
import uuid

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT / "scripts"))
from build import sign, target_key

WINDOWS = os.name == "nt"
SUFFIX = ".exe" if WINDOWS else ""
TARGET = target_key()
DIST = Path(os.environ.get("RCI_DIST", ROOT / "dist")).resolve()
TEMPLATE_TEXT = b'{"version":"0.0.0","marker":"RAFT_NATIVE_FIXTURE_CONFIGURATION_V1"}'
TEMPLATE = TEMPLATE_TEXT + b"\0" * (512 - len(TEMPLATE_TEXT))


def sha(data):
    return hashlib.sha256(data).hexdigest()


def wait_for(predicate, timeout=15):
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        value = predicate()
        if value:
            return value
        time.sleep(0.03)
    raise AssertionError("timed out waiting for the expected process state")


def exchange(home, action="probe"):
    try:
        record = json.loads((home / "fixture-service.json").read_text())
        host, port = record["address"].rsplit(":", 1)
        if host != "127.0.0.1":
            raise AssertionError("fixture address is not loopback")
        with socket.create_connection((host, int(port)), timeout=2) as stream:
            stream.sendall(json.dumps({"action": action, "token": record["token"]}).encode() + b"\n")
            line = stream.makefile("rb").readline(4097)
        result = json.loads(line)
        if result["generation"] != record["generation"]:
            raise AssertionError("fixture generation does not match")
        return {key: result[key] for key in ("pid", "version", "generation")}
    except (OSError, ValueError, KeyError):
        return None


class ReleaseServer:
    def __init__(self):
        self.directory = tempfile.TemporaryDirectory(prefix="rci-releases-")
        self.root = Path(self.directory.name)
        self.releases = {}
        self.channels = {"main": "1.1.0", "alpha": "1.1.0"}
        self.wrong_hash = set()
        self.authority_lies = set()
        self.authority_changes = {}
        self.gzip_versions = set()
        self.manifest_changes = {}
        self.requests = []
        self.request_times = []
        self.slow_checksums = 0.0
        self.tamper_installer = False
        self.missing_checksums = False
        self.channel_resolutions = 0
        self.mutable_redirect = False
        self.machines = []
        self.sidecar = b"isolated native fixture wasm\n"
        self.installer = DIST / "native" / TARGET / ("raft-computer-installer" + SUFFIX)
        self.fixture = DIST / "fixtures" / ("raft-computer-fixture" + SUFFIX)
        self.installer_bytes = self.installer.read_bytes()
        template = self.fixture.read_bytes()
        if template.count(TEMPLATE) != 1:
            raise AssertionError("fixture must contain exactly one configurable native blob")
        self.template = template
        owner = self

        class Handler(http.server.BaseHTTPRequestHandler):
            def do_GET(self):
                owner.respond(self)

            def log_message(self, *_):
                pass

        self.handler = Handler
        self.server = http.server.ThreadingHTTPServer(("127.0.0.1", 0), Handler)
        self.thread = threading.Thread(target=self.server.serve_forever, daemon=True)
        self.thread.start()
        self.base = f"http://127.0.0.1:{self.server.server_port}"
        self.proxies = []

    def publish(self, version, **behavior):
        data = json.dumps({"version": version, **behavior}, separators=(",", ":")).encode()
        if len(data) > 512:
            raise AssertionError("fixture behavior is too large")
        path = self.root / ("fixture-" + version + SUFFIX)
        path.write_bytes(self.template.replace(TEMPLATE, data + b"\0" * (512 - len(data))))
        path.chmod(0o755)
        sign(path)
        self.releases[version] = path.read_bytes()
        return self.releases[version]

    def manifest(self, version):
        data = self.releases[version]
        target = {"file": "raft-computer", "sha256": "0" * 64 if version in self.wrong_hash else sha(data), "size": len(data)}
        manifest = {"version": version, "targets": {TARGET: target},
            "photonWasm": {"file": "photon_rs_bg.wasm", "sha256": sha(self.sidecar), "size": len(self.sidecar)}}
        self.manifest_changes.get(version, lambda _: None)(manifest)
        return manifest

    def respond(self, handler):
        url = urllib.parse.urlsplit(handler.path)
        path = urllib.parse.unquote(url.path)
        query = urllib.parse.parse_qs(url.query)
        self.requests.append(handler.path)
        self.request_times.append((handler.path, time.monotonic()))
        status, body = 200, b""
        parts = path.strip("/").split("/")
        if path == "/public/v2/apps/raft-computer-cli/updates/check":
            version = query.get("version", [self.channels.get(query.get("channel", ["main"])[0])])[0]
            if version not in self.releases:
                status = 404
            else:
                platform, arch = TARGET.split("-")
                # Proxy requests preserve their original URL authority.
                origin = f"{url.scheme}://{url.netloc}" if url.netloc else self.base
                release_id = sha(version.encode())[:32]
                base = f"{origin}/dl/raft-computer-cli/releases/{release_id}/{TARGET}"
                manifest = self.manifest(version)
                raw = manifest["targets"][TARGET]
                wasm = manifest["photonWasm"]
                artifact = {"platform": platform, "arch": arch,
                    "download_url": base if raw["file"] == "raft-computer" else origin + "/outside",
                    "sha256": "0" * 64 if version in self.authority_lies else raw["sha256"], "size_bytes": raw["size"],
                    "photon_wasm": {"download_url": base + "?kind=photon-wasm", "sha256": wasm["sha256"], "size_bytes": wasm["size"]}}
                if version in self.gzip_versions:
                    compressed = gzip.compress(self.releases[version], mtime=0)
                    artifact["gzip"] = {"download_url": base + ".gz", "sha256": sha(compressed), "size_bytes": len(compressed)}
                response = {"update_available": True, "release": {"id": release_id, "version": version}, "artifact": artifact}
                self.authority_changes.get(version, lambda _: None)(response)
                body = json.dumps(response).encode()
        elif len(parts) == 5 and parts[:3] == ["dl", "raft-computer-cli", "releases"]:
            version = next((v for v in self.releases if sha(v.encode())[:32] == parts[3]), None)
            if version not in self.releases or parts[4] not in (TARGET, TARGET + ".gz"):
                status = 404
            elif parts[4].endswith(".gz"):
                if version in self.gzip_versions and not query:
                    body = gzip.compress(self.releases[version], mtime=0)
                else:
                    status = 404
            elif query.get("kind") == ["photon-wasm"]:
                body = self.sidecar
            elif not query:
                body = self.releases[version]
            else:
                status = 404
        elif path == f"/dl/raft-computer-installer/main/{TARGET}":
            self.channel_resolutions += 1
            # A second channel resolution names different bytes. A correct
            # bootstrap freezes this first Location for all subsequent fetches.
            handler.send_response(302)
            location = f"/dl/raft-computer-installer/releases/frozen-{self.channel_resolutions}/{TARGET}"
            if self.mutable_redirect:
                location = f"/dl/raft-computer-installer/alpha/{TARGET}"
            handler.send_header("Location", location)
            handler.end_headers()
            return
        elif path in (f"/dl/raft-computer-installer/releases/frozen-1/{TARGET}", "/installer/SHA256SUMS", f"/installer/native/{TARGET}/raft-computer-installer{SUFFIX}"):
            if path.endswith("SHA256SUMS") or query.get("kind") == ["sha256sums"]:
                time.sleep(self.slow_checksums)
                relative = f"native/{TARGET}/raft-computer-installer{SUFFIX}"
                body = b"" if self.missing_checksums else f"{sha(self.installer_bytes)}  {relative}\n".encode()
            else:
                body = self.installer_bytes + (b"tampered" if self.tamper_installer else b"")
        else:
            status = 404
        handler.send_response(status)
        handler.send_header("Content-Length", str(len(body)))
        handler.end_headers()
        try:
            handler.wfile.write(body)
        except (BrokenPipeError, ConnectionResetError):
            pass

    def proxy(self):
        server = http.server.ThreadingHTTPServer(("127.0.0.1", 0), self.handler)
        thread = threading.Thread(target=server.serve_forever, daemon=True)
        thread.start()
        self.proxies.append((server, thread))
        return f"http://127.0.0.1:{server.server_port}"

    def machine(self):
        machine = Machine(self)
        self.machines.append(machine)
        return machine

    def close(self):
        for machine in self.machines:
            machine.close()
        for server, thread in [*self.proxies, (self.server, self.thread)]:
            server.shutdown()
            server.server_close()
            thread.join(timeout=5)
        self.directory.cleanup()


class Machine:
    def __init__(self, server, prefix="rci-machine-"):
        self.server = server
        self.directory = tempfile.TemporaryDirectory(prefix=prefix)
        self.home = Path(self.directory.name).resolve()
        self.install_dir = self.home / "bin"
        self.install_dir.mkdir()
        self.k = self.home / "computer" / "k"
        self.state = self.home / "computer" / "installer"

    @property
    def binary(self):
        return self.install_dir / ("raft-computer" + SUFFIX)

    def env(self, extra=None):
        environment = {key: value for key, value in os.environ.items() if key.lower() not in (
            "http_proxy", "https_proxy", "all_proxy", "no_proxy") and not key.startswith(("RAFT_", "SLOCK_", "RCI_FIXTURE_"))}
        environment.update(HOME=str(self.home), USERPROFILE=str(self.home), SHELL="/bin/zsh", RAFT_HOME=str(self.home),
            RAFT_COMPUTER_INSTALL_DIR=str(self.install_dir), RAFT_COMPUTER_RELEASE_BASE=self.server.base + "/computer",
            RAFT_COMPUTER_HANDS_ORIGIN=self.server.base, RAFT_COMPUTER_NON_INTERACTIVE="1", RAFT_COMPUTER_NO_MODIFY_PATH="1" if WINDOWS else "0",
            RAFT_COMPUTER_INSTALLER_DL_BASE=self.server.base + "/dl/raft-computer-installer")
        environment.update(extra or {})
        return environment

    def command(self, args, extra=None, bootstrap=False, psreadline=False):
        if bootstrap:
            if WINDOWS:
                powershell = str(Path(os.environ["SystemRoot"]) / "System32/WindowsPowerShell/v1.0/powershell.exe")
                if psreadline:
                    # An interactive Windows PowerShell 5.1 console has PSReadLine
                    # loaded before the entry script runs. Reproduce that state
                    # explicitly: -File cannot, because it starts a bare session.
                    quoted = " ".join("'" + a.replace("'", "''") + "'" for a in args)
                    script = "Import-Module PSReadLine; & '" + str(DIST / "install.ps1") + "' " + quoted + "; exit $LASTEXITCODE"
                    return [powershell, "-NoProfile", "-ExecutionPolicy", "Bypass", "-Command", script]
                return [powershell, "-NoProfile", "-ExecutionPolicy", "Bypass", "-File", str(DIST / "install.ps1"), *args]
            return ["/bin/sh", str(DIST / "install.sh"), *args]
        return [str(self.server.installer), *args]

    def run(self, args, extra=None, bootstrap=False, psreadline=False):
        return subprocess.run(self.command(args, extra, bootstrap, psreadline), env=self.env(extra), text=True, capture_output=True, timeout=180)

    def json(self, args, expected=0, extra=None):
        result = self.run([*args, "--json"], extra)
        assert result.returncode == expected, (result.returncode, result.stdout, result.stderr)
        reply = json.loads(result.stdout)
        assert reply["exitCode"] == expected
        return reply

    def product(self, args):
        result = subprocess.run([str(self.binary), *args], env=self.env(), text=True, capture_output=True, timeout=20)
        assert result.returncode == 0, (result.stdout, result.stderr)
        return result

    def self_version(self):
        return self.product(["--version"]).stdout.strip()

    def live(self):
        return exchange(self.home)

    def login_start(self):
        self.product(["login"])
        self.product(["start"])
        return wait_for(self.live)

    def preinstall(self, version, running=False):
        self.binary.write_bytes(self.server.releases[version])
        self.binary.chmod(0o755)
        (self.install_dir / "photon_rs_bg.wasm").write_bytes(self.server.sidecar)
        if running:
            self.login_start()

    def receipt(self, id):
        return json.loads((self.state / "receipts" / (sha(id.encode()) + ".json")).read_text())

    def close(self):
        exchange(self.home, "stop")
        deadline = time.monotonic() + 5
        while self.live() and time.monotonic() < deadline:
            time.sleep(0.05)
        if self.live():
            raise AssertionError("fixture service did not stop during cleanup")
        self.directory.cleanup()
